src/plugins/hardware-models/hardware_models_thread.cpp

```diff
@@ -0,0 +1,348 @@
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Library General Public License for more details.
+ *
+ *  Read the full text in the LICENSE.GPL file in the doc directory.
+ */
+
+#include "hardware_models_thread.h"
+
+#include <config/yaml.h>
+#include <core/threading/mutex_locker.h>
+#include <interfaces/SwitchInterface.h>
+#include <utils/misc/map_skill.h>
+#include <utils/misc/string_conversions.h>
+#include <utils/misc/string_split.h>
+#include <utils/time/wait.h>
+
+#include <unistd.h>
+
+using namespace fawkes;
+
+/** @class HardwareModelsThread "hardware_models_thread.h"
+ * Main thread of Hardware Models Plugin.
+ *
+ * Parses yaml files that are assumend to have the following format:
+ * COMPONENT-NAME:
+ *  states:
+ *      \verbatim<List of states, the first state is assumed to be the initial state>\endverbatim
+ * 
+ *  STATE_1:
+ *   edges:
+ *    \verbatim<List of states the state has an edge to>\endverbatim
+ *   edge_1:
+ *    transition: \verbatim<name of the action that causes the transition>\endverbatim
+ *    (optional)probability: \verbatim<probability of the transition if it is an exogenous action>\endverbatim
+ *     
+ *  Any component plugin can inform the HardwareModel Plugin about state changes
+ *  by sending a HardwareModelInterfaceMessage.
+ * @author Daniel Habering
+ */
+
+/** Constructor. */
+HardwareModelsThread::HardwareModelsThread()
+: Thread("HardwareModelsThread", Thread::OPMODE_WAITFORWAKEUP),
+  CLIPSFeature("hardware-models"),
+  CLIPSFeatureAspect(this),
+  BlackBoardInterfaceListener("HardwareModelsThread")
+
+{
+}
+
+void
+HardwareModelsThread::init()
+{
+	std::string cfg_interface_ = "HardwareModels";
+	if (config->exists("/hardware-models/interface")) {
+		cfg_interface_ = config->get_string("/hardware-models/interface");
+	}
+
+	hm_if_ = blackboard->open_for_writing<HardwareModelsInterface>(cfg_interface_.c_str());
+	blackboard->register_listener(this);
+	wakeup();
+	bbil_add_message_interface(hm_if_);
+}
+
+/**
+ * @brief Initializes hardware components from yaml files for a clips environment
+ * 
+ * @param env_name Name of clips environment
+ * @param clips Pointer to clips environment
+ */
+void
+HardwareModelsThread::clips_context_init(const std::string &          env_name,
+                                         LockPtr<CLIPS::Environment> &clips)
+{
+	std::string models_dir_;
+
+	models_dir_ = config->get_string("/hardware-models/models-dir");
+	if (models_dir_[models_dir_.size() - 1] != '/') {
+		models_dir_ += "/";
+	}
+
+	envs_[env_name] = clips;
+
+	clips->batch_evaluate(SRCDIR "/hardware_models.clp");
+
+	components_ = config->get_strings("/hardware-models/components");
```

Comment:
This is actually not needed at all. The plugins assumes that there is a config value /hardware-models/components somewhere (presumably in conf.d/hardware-models.yaml. This then defines a list of components, for which a tree of config values should be defined (As shown in the example file gripper.yaml)
Where this is defined, does not really matter, since the config parser takes care of parsing all relevant config files.

However I have removed the models_dir_ variable